src/audio_unit/delay.rs

```diff
@@ -1,32 +1,102 @@
 use crate::{ring_buffer, AudioUnit, Result};
+use anyhow::anyhow;
 use cpal::StreamConfig;
 use ringbuf::{Consumer, Producer, RingBuffer};
+use wmidi::MidiMessage;
+
+pub type DelayMs = u32;
+
+type MidiMessageHandler = Box<dyn Fn(&[MidiMessage<'static>]) -> Option<DelayMs> + Send + Sync>;
 
 pub struct Delay {
+    delay_ms: DelayMs,
+    midi_message_handler: MidiMessageHandler,
+    stream_config: StreamConfig,
     producer: Producer<f32>,
     consumer: Consumer<f32>,
 }
 
 impl Delay {
-    pub fn new(stream_config: &StreamConfig, delay_ms: u32) -> Result<Self> {
-        let delay_num_frames = (delay_ms as f32 / 1_000.0) * stream_config.sample_rate.0 as f32;
-        let delay_num_samples = delay_num_frames as usize * stream_config.channels as usize;
+    pub const MIN_DELAY_MS: DelayMs = 0;
+    pub const MAX_DELAY_MS: DelayMs = 10000;
```

Comment:
We need some upper bound here because the ring buffer isn't resizable. (Also something to map the MIDI input to.)